### ISSUE TITLE
Photos: tagging and untagging when editing photos broken

### ISSUE THREAD

**Issue by AlfredSK** [friendica/friendica#5915]
### Expected behavior
When editing a photo I can add and remove tags to or from the photo. 
Adding tags should work by adding a list of tags like #tag1, #tag2, #tag3.
When I try to remove all the tags from the photo (all checkboxes set) all tags should be removed at once.

### Actual behavior
When I try to add like three tags to a photo not all tags are added to the photo but only two of them. I can add more tags afterwards. But some tags are missing when adding multiple tags at once.
When removing all the tags at once (all checkboxes set) only one tag is removed. I have to repeat the process for every tag. The last tag cannot be removed at all.

### Steps to reproduce the problem
Go to a photo you uploaded and edit it. URL looks like this...
.../photos/user/image/12345....12345/edit
Try to add a list of tags like #tag1, #tag2, #tag3 and save your edit.
Try to remove all the tags by clicking on the 'tag remove' link. Select all the checkboxes and click on remove.

### Friendica version you encountered the problem
2018.12-dev

### Friendica source (git, zip)
git

### PHP version
7.2.10

### SQL version
10.1.34 MariaDB

**Comment by JonnyTischbein** [comment]
I'm a bit confused. Can anyone please help me to understand where the tags are stored ?

So far I know when displaying an image:
- The link contains the `resource-id` in `item`
- via this I get the `id`
- the field `tag` should contain the tags
  - BUT: If I select it with mysql it is displayed as `NULL`
  - Via friendica the content is: 

        #[url=https://friendica.riotcat.org/search?tag=tag]tag[/url],#[url=https://friendica.riotcat.org/search?tag=tag3]tag3[/url]

(I was looking into file `mod/tagrm` and `mod/photos`)

Thanks!

**Comment by MrPetovan** [comment]
Look in the `term` table with `oid = 12345` where 12345 is the item ID used in `mod/tagrm`,

The `Item` model pretends you can query the deprecated `item.tag` field when it actually performs a `Term::tagTextFromItemId()` calls in `Model\Item` line 226.

**Comment by JonnyTischbein** [comment]
Maybe this is more a feature request than fixing this issue but wouldn't a way like removing the categories from a post a better solution like the `/tagrm` page?

In both cases you need click at every tag.

![screenshot_2018-10-22_19-49-58](https://user-images.githubusercontent.com/10757520/47309058-bb5e5d80-d633-11e8-92d2-531d5c40eef7.png)

**Comment by MrPetovan** [comment]
Yes, definitely.

**Comment by JonnyTischbein** [comment]
I'll need again help with the `Item::update()` and the `tag` field in item.

The problem is that if I want to remove every tag the array will be empty, because every variable in the array was unset.

So in this case  `$tag_str = ''`
    
    /mod/tagrm.php:60 
    Item::update(['tag' => $tag_str], ['id' => $item_id]);

But in `Item::update`.   Field tag is set, but `$fields['tag'] = $tags = null;`

    /src/Model/Item.php:815
    public static function update(array $fields, array $condition)
    [...]
    if (array_key_exists('tag', $fields)) {
		$tags = $fields['tag'];
		$fields['tag'] = null;
	} else {
		$tags = '';
	}

So the update will not be performed.

    /src/Model/Item.php:898
    if (!empty($tags)) {
        Term::insertFromTagFieldByItemId($item['id'], $tags);
         if (!empty($item['tag'])) {
             DBA::update('item', ['tag' => ''], ['id' => $item['id']]);
         }
     }

I wanted to use the `Item::delete` function, but there is no way to provide the `tag` field to select the `term` table.
Is there a way to update the `tag` "field" to null, do I need to change the `Ìtem.php` ?

**Comment by annando** [comment]
Possibly it is an idea to not check for empty but to check for existence of the array element like it is done on other places in that code.

**Comment by MrPetovan** [comment]
The usual way of knowing is a field has been emptied out is the following:
- If tags should be untouched, `$tags` will be `null`.
- If tags have been changed, `$tags` will be set.
- If tags have been cleared, `$tags` will be an empty string.

In pseudo-code:
```php
if (array_key_exists('tag', $fields)) {
	$tags = $fields['tag'];
	$fields['tag'] = null;
} else {
	$tags = null;
}

if (is_null($tags)) {
	// Delete tags, maybe with a method from Term
	//Term::delete...
	
	if ($tags) {
		Term::insertFromTagFieldByItemId($item['id'], $tags);
		if (!empty($item['tag'])) {
			DBA::update('item', ['tag' => ''], ['id' => $item['id']]);
		}
	}
}
```

**Comment by MrPetovan** [comment]
This issue needs testing after #6003 has been merged.

**Comment by JonnyTischbein** [comment]
Things I tested:
- adding 4-5 one words tags like `#a #b #c #d` or `#a, #b, #c, #d`
- removing them all by clicking the `(x)` (using the GET method `/tagrm/ID/TAG`)
- removing some but not all tags via `[Select tags to remove]`
- removing all tags via `[Select tags to remove]`

The removing all or removing the last tag was the part didn't working before.

**Comment by AlfredSK** [comment]
If I may make a wish, I'd like to have some space between the colored tags. :relaxed:

**Comment by rabuzarus** [comment]
One question: Why have you labeled the photo tags as `category` (https://github.com/friendica/friendica/pull/6003/files#diff-579cb027398856b2c2af6e045b0bbd13R56) and not as tag (`<span class="tag label btn-info sm">`)?

**Comment by JonnyTischbein** [comment]
Sorry. I was using categories from post as template

**Comment by AlfredSK** [comment]
Fixed by #6003 

Closing.
